vms/platformvm/state/stakers_model_generator.go

```diff
@@ -0,0 +1,212 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package state
+
+import (
+	"reflect"
+	"time"
+
+	blst "github.com/supranational/blst/bindings/go"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+	"github.com/ava-labs/avalanchego/vms/platformvm/txs"
+	"github.com/leanovate/gopter"
+	"github.com/leanovate/gopter/gen"
+)
+
+type generatorPriorityType uint8
+
+const (
+	anyPriority generatorPriorityType = iota
+	currentValidator
+	currentDelegator
+	pendingValidator
+	pendingDelegator
+)
+
+// stakerGenerator helps creating random yet reproducible Staker objects,
+// which can be used in our property tests. stakerGenerator takes care of
+// enforcing some Staker invariants on each and every random sample.
+// TestGeneratedStakersValidity documents and verifies the enforced invariants.
+func stakerGenerator(
+	prio generatorPriorityType,
+	subnet *ids.ID,
+	nodeID *ids.NodeID,
+	maxWeight uint64, // helps avoiding overflows in delegator tests
+) gopter.Gen {
+	return genStakerTimeData(prio).FlatMap(
+		func(v interface{}) gopter.Gen {
+			macro := v.(stakerTimeData)
+
+			genStakerSubnetID := genID
+			genStakerNodeID := genNodeID
+			if subnet != nil {
+				genStakerSubnetID = gen.Const(*subnet)
+			}
+			if nodeID != nil {
+				genStakerNodeID = gen.Const(*nodeID)
+			}
+
+			return gen.Struct(reflect.TypeOf(Staker{}), map[string]gopter.Gen{
+				"TxID":            genID,
+				"NodeID":          genStakerNodeID,
+				"PublicKey":       genBlsKey,
+				"SubnetID":        genStakerSubnetID,
+				"Weight":          gen.UInt64Range(0, maxWeight),
+				"StartTime":       gen.Const(macro.StartTime),
+				"EndTime":         gen.Const(macro.EndTime),
+				"PotentialReward": gen.UInt64(),
+				"NextTime":        gen.Const(macro.NextTime),
+				"Priority":        gen.Const(macro.Priority),
+			})
+		},
+		reflect.TypeOf(stakerTimeData{}),
+	)
+}
+
+// stakerTimeData holds Staker's time related data in order to generate them
+// while fullfilling the following constrains:
```

Comment:
Done, thanks

> would it be possible to move all the property based tests to a separate dir?

They used to be in a different package. What is annoying is that we're going to use to the tests the staker section of state package. If properties are in a different package (and we keep stakers in state) we won't get the test coverage, which was very helpful to me to spot untested code path and duly cover them or simplify them.

I guess the best solution would be to repackage the stakers section of state and their properties together. I am in favour of it but I'd do that in a different PR (also the state.load section will get slightly more complex since we load staker from genesis in an ad-hoc way)